### PR TITLE
Visualization action points

### PR DESCRIPTION
**Pull request by ChrnyaevEK**
This is a merge request for AP visualization, as we have discussed.

Basically I have changed `ActionStateBefore` event scheme, then I have updated runtime `@action` decorator to get AP ids from arguments. Due to this change the `ActionStateBefore` will be dispatched for each action (with or without id or any pose/joints in arguments). This event will be than handled by rest proxy. 

At rest proxy I have added 2 global variables to hold last action state - one for `ActionStateBefore.Data` and another for `ActionStateAfter.Data`, just like you hold `package_state` for example. `ActionStateAfter` will be unset once `ActionStateBefore` is met (at this moment `ActionStateAfter` already belongs to previous action), however `ActionStateBefore` will never be unsettled once set. We have discussed internally, that it is ok, that information about the last AP will remain (will not be unset), even when script state is "Completed" or action has ended (`ActionStateAfter` was fired). `ExecutionInfo` and `packeges/executioninfo` were updated to return AP ids.

As well I have updated tests for `arcor_runtime`, but I guess you might want to check some other things here, so it is an open question what tests should be done.

I ran tests for related services. I did not manage to run typecheck and tests for all services due to (I suppose) `pyk4a`.

### CODE REVIEW

File: src/python/arcor2_execution_rest_proxy/scripts/execution_rest_proxy.py [hunk 1]
```diff
@@ -920,6 +935,9 @@ def packages_executioninfo() -> RespT:
     else:
         ret = ExecutionInfo(ExecutionState.Undefined)  # TODO this is unhandled state - log it
 
+    if action_state_before is not None:
+        ret.actionPointIds = action_state_before.action_point_ids
+
```
Comment (ChrnyaevEK) [inline src/python/arcor2_execution_rest_proxy/scripts/execution_rest_proxy.py:940]: I tried to use `try...except` here, but type checking failed with "None" has no attribute...

File: src/python/arcor2_runtime/action.py [hunk 2]
```diff
@@ -173,32 +172,34 @@ def wrapper(*args: Union[Generic, Any], an: Optional[str] = None, **kwargs: Any)
...
-                except AttributeError:
+            if action_id and _executed_action is None:
+                _executed_action = action_id, f
+
+        # collect action point ids, ignore missing
+        action_point_ids = []
+        for aa in action_args:
+            try:
+                if isinstance(aa, Pose) or isinstance(aa, ProjectRobotJoints):
+                    action_point_ids.append(getattr(aa, AP_ID_ATTR))
```
Comment (ChrnyaevEK) [inline src/python/arcor2_runtime/action.py:183]: Should we check other types, like Orientation or Position?

File: src/python/arcor2_execution_rest_proxy/scripts/execution_rest_proxy.py (hunk 1 above)
Comment (ZdenekM) [reply src/python/arcor2_execution_rest_proxy/scripts/execution_rest_proxy.py:940]: You mean `try...except` to catch what? I guess it should not be necessary here, unless I missed something :-) Btw, you can simplify `if action_state_before is not None:` to just `if action_state_before:` here safely.

File: src/python/arcor2/data/events.py [hunk 3]
```diff
@@ -118,9 +118,15 @@ class ActionStateBefore(Event):
     @dataclass
     class Data(JsonSchemaMixin):
 
-        action_id: str
+        # required for projects with logic (where actions are defined in the project)
+        # might be also filled for projects without logic
+        action_id: Optional[str]
```
Comment (ZdenekM) [inline src/python/arcor2/data/events.py:123]: Please add a default value (`=None`).

File: src/python/arcor2/data/events.py [hunk 4]
```diff
@@ -118,9 +118,15 @@ class ActionStateBefore(Event):
...
         parameters: List[str]
```
Comment (ZdenekM) [inline src/python/arcor2/data/events.py:124]: This should be also optional and have a default value, e.g. `parameters: Optional[List[str]] = None`.

File: src/python/arcor2/data/events.py [hunk 5]
```diff
@@ -118,9 +118,15 @@ class ActionStateBefore(Event):
...
 
+        # required for projects with logic
+        # might or might not be set in other cases
+        action_point_ids: List[str]
```
Comment (ZdenekM) [inline src/python/arcor2/data/events.py:128]: Make it optional with a default value.

File: src/python/arcor2_runtime/action.py [hunk 6]
```diff
@@ -173,32 +172,34 @@ def wrapper(*args: Union[Generic, Any], an: Optional[str] = None, **kwargs: Any)
...
-                        isinstance(aa, ProjectRobotJoints) and getattr(aa, AP_ID_ATTR) in breakpoints
-                    ):
-                        make_a_break = True
-                        break
-                except AttributeError:
+            if action_id and _executed_action is None:
+                _executed_action = action_id, f
+
+        # collect action point ids, ignore missing
+        action_point_ids = []
```
Comment (ZdenekM) [inline src/python/arcor2_runtime/action.py:179]: Please add type annotation.

File: src/python/arcor2_runtime/action.py (hunk 2 above)
Comment (ZdenekM) [reply src/python/arcor2_runtime/action.py:183]: I think that it is ok to check just `Pose` (actually its `position` property, see the original code) and `ProjectRobotJoints` - only these two types are patched with the `AP_ID_ATTR`. We don't support (at the moment) parameters of type `Orientation` or `Position` anyway.

File: src/python/arcor2_runtime/action.py [hunk 7]
```diff
@@ -173,32 +172,34 @@ def wrapper(*args: Union[Generic, Any], an: Optional[str] = None, **kwargs: Any)
...
+                if isinstance(aa, Pose) or isinstance(aa, ProjectRobotJoints):
+                    action_point_ids.append(getattr(aa, AP_ID_ATTR))
+            except AttributeError:
+                if breakpoints:
                     raise Arcor2Exception("Orientations/Joints not patched. Breakpoints won't work!")
 
+        # validate if break is required
+        make_a_break = bool(breakpoints and breakpoints.intersection(action_point_ids))
+
+        # dispatch even if action id was not detected
```
Comment (ZdenekM) [inline src/python/arcor2_runtime/action.py:191]: dispatch even**t** :)

**Review by ZdenekM — CHANGES_REQUESTED** [review]
Thanks for your effort, it certainly looks good.

**Comment by ZdenekM** [comment]
If you have problems with `pyk4a`, get rid of `arcor2_kinect_azure`, remove `pyk4a` from `requirements.txt` and you should be fine. Another option is to install it (we have a [script](https://github.com/robofit/arcor2/blob/master/build-support/install_kinect_prerequisites.sh) for that, certainly works on Ubuntu 18.04).

Regarding tests, I guess that it will be also necessary to update [arcor2_arserver/tests/test_project_execution.py](https://github.com/robofit/arcor2/blob/master/src/python/arcor2_arserver/tests/test_project_execution.py).

And the last thing - please squash your commits and follow https://www.conventionalcommits.org/en/v1.0.0/.

**Comment by ZdenekM** [comment]
@ChrnyaevEK Sorry for the trouble, in the meantime, I've made some changes regarding switching to Python 3.9, which caused some conflicts... Please rebase your branch.

File: src/python/arcor2/data/events.py [hunk 8]
```diff
@@ -118,8 +118,14 @@ class ActionStateBefore(Event):
...
-        action_id: str
-        parameters: List[str]
+        # required for projects with logic (where actions are defined in the project)
+        # might be also filled for projects without logic
+        action_id: Optional[str] = None
+        parameters: Optional[List[str]] = None
+
+        # required for projects with logic
+        # might or might not be set in other cases
+        action_point_ids: Optional[List[str]] = None
```
[the remark this answers is not in the document]
Comment (ZdenekM) [reply src/python/arcor2/data/events.py:128]: Well, actually `set[str]` after the switch to 3.9...

File: src/python/arcor2_execution_rest_proxy/scripts/execution_rest_proxy.py (hunk 1 above)
Comment (ChrnyaevEK) [reply src/python/arcor2_execution_rest_proxy/scripts/execution_rest_proxy.py:940]: Well, yes, [this](https://docs.python.org/3.6/glossary.html#term-eafp) is not really necessary here, agree. But I would leave `if action_state_before is not None` as it is recommended to use `is` when comparing with `None`  ([PEP 8](https://www.python.org/dev/peps/pep-0008/#programming-recommendations))

File: src/python/arcor2/data/events.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ChrnyaevEK) [reply src/python/arcor2/data/events.py:128]: Agree. Will be updated with next commit

**Comment by ZdenekM** [comment]
@ChrnyaevEK Tests failed due to some rare race condition (not because of your changes), that should be hopefully fixed in https://github.com/robofit/arcor2/pull/708.

**Comment by ChrnyaevEK** [comment]
@ZdenekM Thanks! At least I have installed pyk4a to check it locally and locally it succeeded

**Review by ZdenekM — APPROVED** [review]

**Pull request merged**